File: hackerrank-java/MaxSubarray.java
public class MaxSubarray {
    public static void main(String[] args) {
        
    }

    public int maxSubArray(int[] nums) {
        int currSum = nums[0];
        for (int i = 0; i < nums.length; i++) {
            int sum = 0;
            for (int j = i; j < nums.length; j++) {
                sum += nums[j];
                if (sum > currSum) {
                currSum = sum;
            }
            }
        }
        return currSum;
        
    }

    public int maxSubArray2(int[] nums) {
        int global = nums[0];
        int local = nums[0];
        for (int i = 1; i < nums.length; i++) {
            local += nums[i];
            global = Math.max(local, global);
            if (local < 0) {
                local = 0;
            }
        }
        return global;
    }
}

File: CrackingTheCodingInterview/MaxSubArray.js
var maxSubArray = function(nums) {
    let local = nums[0];
    let global = nums[0];
    for (let i = 0; i < nums.length; i++) {
        local = Math.max(nums[i], nums[i] + local);
        global = Math.max(local, global);
    }
    return global;
};